core/src/main/java/org/elasticsearch/index/shard/RefreshListeners.java

```diff
@@ -0,0 +1,156 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.shard;
+
+import org.apache.lucene.search.ReferenceManager;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.engine.Engine;
+import org.elasticsearch.index.translog.Translog;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.concurrent.LinkedTransferQueue;
+import java.util.function.Consumer;
+import java.util.function.IntSupplier;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Allows for the registration of listeners that are called when a change becomes visible for search. This functionality is exposed from
+ * {@link IndexShard} but kept here so it can be tested without standing up the entire thing. 
+ */
+class RefreshListeners {
+    /**
+     * Refresh listeners. While they are not stored in sorted order they are processed as though they are.
+     */
+    private final LinkedTransferQueue<Tuple<Translog.Location, Consumer<Boolean>>> refreshListeners = new LinkedTransferQueue<>();
+
+    private final IntSupplier getMaxRefreshListeners;
+    private final Runnable forceRefresh;
+    private final Consumer<Runnable> fireListener;
+
+    /**
+     * The estimated size of refreshListenersEstimatedSize used for triggering refresh when the size gets over
+     * {@link IndexSettings#MAX_REFRESH_LISTENERS_PER_SHARD}. No effort is made to correct for threading issues in the size calculation
+     * beyond it being volatile.
+     */
+    private volatile int refreshListenersEstimatedSize;
+    /**
+     * The translog location that was last made visible by a refresh.
+     */
+    private volatile Translog.Location lastRefreshedLocation;
+
+    public RefreshListeners(IntSupplier getMaxRefreshListeners, Runnable forceRefresh, Consumer<Runnable> fireListener) {
+        this.getMaxRefreshListeners = getMaxRefreshListeners;
+        this.forceRefresh = forceRefresh;
+        this.fireListener = fireListener;
+    }
+
+    /**
+     * Add a listener for refreshes.
+     *
+     * @param location the location to listen for
+     * @param listener for the refresh. Called with true if registering the listener ran it out of slots and forced a refresh. Called with
+     *        false otherwise.
+     */
+    public void add(Translog.Location location, Consumer<Boolean> listener) {
+        requireNonNull(listener, "listener cannot be null");
+        Translog.Location listenerLocation = requireNonNull(location, "location cannot be null");
+
+        Translog.Location lastRefresh = lastRefreshedLocation;
+        if (lastRefresh != null && lastRefresh.compareTo(listenerLocation) >= 0) {
```

Comment:
I usually prefer this style. Is this code special in a way that needs the other?